### ISSUE TITLE
Bone color should be ossified

### ISSUE THREAD

**Issue by tkuchida** [opensim-org/opensim-gui#517]
Current colors make bones look metallic:
![](https://user-images.githubusercontent.com/4203505/36235242-41afde2c-11a4-11e8-83bb-e16bc908eaf0.png)

Appearance depends on lighting conditions, but something like this might be better (RGB = 242, 220, 193):
![](https://user-images.githubusercontent.com/4203505/36235393-29033f44-11a5-11e8-8c53-31ed881b16c7.png)

Using OpenSim-078b8daf-2018-01-26.zip from AppVeyor (with no model spotlight and with ambient light intensity increased from 0.2 to 0.4 in OpenSimEditor.js).

We could change the colors in the model files, but ideally we would have the visualizer equivalent of a CSS file and simply specify "bone" as the color in the model file.

@aseth1

**Comment by aymanhab** [comment]
@tkuchida Nice selection of colors. My thought though is that color and texture are two different things. Texture (e.g. ossified) would be related to shininess, metalness, maybe a Map. We should provide a built in list of these that users can refer to by name. Color is a well-defined coat of paint that can be applied to any texture. Would be good to come up with bone "texture" and if you like the bones with RGB values above you still can use them. Thoughts?

**Comment by tkuchida** [comment]
> My thought though is that color and texture are two different things. Texture (e.g. ossified) would be related to shininess, metalness, maybe a Map. We should provide a built in list of these that users can refer to by name. Color is a well-defined coat of paint that can be applied to any texture. Would be good to come up with bone "texture" and if you like the bones with RGB values above you still can use them. Thoughts?

Yes, I know there is a difference between color and texture. I was not proposing that we texture map the bones, but I'm sure that would look great if done well. I've changed the title of the Issue to avoid the incorrect interpretation of the word "ossified" to imply changing the texture. I think changing the color from pure 255,255,255 white would go a long way with minimal work.

**Comment by chrisdembia** [comment]
I think @aymanhab is using "texture" in a more general sense, suggesting not necessarily using a texture map, but simply changing the material properties of the bone.

Changing the default color seems like the obvious thing to do, but if modifying shininess etc. is as easy and can produce as nice of an effect as the second screenshot above, then that also sounds fine.

**Comment by tkuchida** [comment]
@chrisdembia I have changed the title to "Bones of model do not look like human bones", which is the underlying issue. Perhaps we need to discuss how to fix it and who has time to work on it.

**Comment by aseth1** [comment]
Wasn't one of the arguments for the creation of `Appearance` was to combine visual attributes like color, texture, opacity to create a list of predefined appearances to represent materials (like steel, bone, chrome, ...) as well as standardize the look of wrap and contact surfaces? Describing a *Bone* `Appearance` would appear to be the perfect use case.

@aymanhab where should these predefined Appearances live? Now that an actual `Appearance` is a property of `Geometry`, how do you suggest we populate Geometry with a preset Appearance? Also would we then have `Geometry::Mesh` have a default appearance of Bone?

**Comment by aymanhab** [comment]
@aseth1  as of now we don't have an abstraction for Material and every geometry  has Appearance Properties. We can use the "texture" Property to control attributes like "Roughness", "Shininess", ... for any set of Geometries. The mapping would be hardcoded in the GUI, or in json files that the visualizer loads directly.
@tkuchida Bones in three.js have no rendering attributes, they are basically  the control points of meshes when doing skinning.

**Comment by chrisdembia** [comment]
> Also would we then have Geometry::Mesh have a default appearance of Bone?

The property system does not yet have a formal way of a subclass having a different default value for a property in a base class.

**Comment by aymanhab** [comment]
Chris, we would just change model files by adding texture string, some
geimetry have bone other metal, etc.

**Comment by tkuchida** [comment]
@aymanhab I've played around with the settings in Expert Mode using the localhost address trick in Chrome, but was unable to get anything better than just changing the mesh color.

| Original | Bones changed to 0.949/0.863/0.757 |
| --- | --- |
| ![](https://user-images.githubusercontent.com/4203505/38396949-cd10a3bc-38ef-11e8-8bd3-d688e384c26b.png) | ![](https://user-images.githubusercontent.com/4203505/38396955-d3c37ca2-38ef-11e8-874f-4c3cfa9cb1a7.png) |

**Comment by tkuchida** [comment]
@aymanhab I've tried using the Expert Mode trick to change properties of the mesh but I'm getting mixed results and it isn't clear whether I'm barking up the wrong tree or whether the tree is malfunctioning. Here's the original:

![](https://user-images.githubusercontent.com/4203505/38830706-0cc88468-4172-11e8-8014-e3c374c88d85.png)

The skull looks shiny/metallic. I'm able to change the color:

![](https://user-images.githubusercontent.com/4203505/38830736-20747404-4172-11e8-8c57-786f6f11c7c9.png)

but it still looks metallic. My guess is that the shininess is caused by the "shininess" property, which is currently set to the default of 30% for MeshPhongMaterial. I can change the property in Expert Mode but it has no effect on the visualization.

**Shininess = 0%:**
![](https://user-images.githubusercontent.com/4203505/38830843-639cca4c-4172-11e8-812b-bbe6f690ff40.png)

**Shininess = 100%:**
![](https://user-images.githubusercontent.com/4203505/38830847-67bf5c84-4172-11e8-9b2a-288c63b62026.png)

I think there are two possible explanations: either my hypothesis is incorrect and the apparent shininess is not due to the 30% shininess of the mesh, or my hypothesis is correct and the visualizer isn't being updated when I change the shininess property in Expert Mode. I have not been able to build the GUI on my machine and I haven't figured out a way to hack the javascript to change the shininess property (the `materials` array is indexed by UUID so I can't just cycle through the array to update the property of all meshes). Would you please investigate or provide me with some direction as to what I should do to proceed? I think the bones would look better if we can remove the specular highlights. Thanks.

**Comment by aymanhab** [comment]
@tkuchida I tried a few things, let's discuss in the morning.

**Comment by tkuchida** [comment]
@aymanhab The bone material looks much better now (as of AppVeyor artifact OpenSim-e5f97340-2018-04-20.zip). I've made the following additional tweaks on my end:
- Changed the `sceneLightColor` (the color used for `directionalLight`) from `12040119` (the decimal equivalent of `0xb7b7b7`) to `0xf0f0f0` [in OpenSimEditor.js].
- Changed color of bone meshes to `RGB(237, 215, 198)` [in the model file].
- Changed muscle color functions to avoid purple shades in mid range.

![](https://user-images.githubusercontent.com/4203505/39608970-239a5546-4efa-11e8-9092-d1f5f6e0aec9.png)

**Comment by aymanhab** [comment]
Love the colors indeed. Great job!

**Comment by aseth1** [comment]
I would also add that bones look more jagged/aliased and less smooth in new visualizer than in 3.3:
![image](https://user-images.githubusercontent.com/6494122/40211808-92dcff28-5a01-11e8-9ae3-816e3e28d166.png)
Versus  this in 3.3 -
![image](https://user-images.githubusercontent.com/6494122/40211848-be7cbe2a-5a01-11e8-900c-6588ea8154e3.png)

**Comment by tkuchida** [comment]
Can be improved somewhat with better lighting, but yes it looks like 3.3 smoothed the mesh or surface normals or something.

![](https://user-images.githubusercontent.com/4203505/40212715-451b648c-5a06-11e8-81ac-fea62fa0bc49.png)

<pre>
SunLight position (-10,-5,1)
SunLight intensity 0.3
AmbientLight intensity 0.22
CameraLight intensity 0.7
</pre>

**Comment by tkuchida** [comment]
@aseth1 There needs to be a call to `geometry.computeVertexNormals()` after constructing the mesh geometry. @aymanhab I've hacked it in locally at the bottom of OpenSimLoader.js, before returning `object`:

| **Before** | **After** |
| --- | --- |
| ![](https://user-images.githubusercontent.com/4203505/40218063-70d193d0-5a24-11e8-8963-a8e665004ac5.png) | ![](https://user-images.githubusercontent.com/4203505/40218069-74e6876e-5a24-11e8-98fc-edb84dbd25fe.png) |

**Comment by chrisdembia** [comment]
Oh my gosh that's amazing. So much better.

**Comment by aymanhab** [comment]
Definitely better. I remember setting this flag years ago and never
touching it again. Maybe we didn't need that with Phing material.
Definitely will revisit. Thanks.

**Comment by tkuchida** [comment]
Please switch the flag on. We may need to dial roughness back a bit now... I'll investigate with David and follow up.

**Comment by aymanhab** [comment]
Sure will, and please keep me posted in material settings. Will we do with
colors in meshes in v3.3 format since texture upgrade is substantial task?

**Comment by tkuchida** [comment]
Not fixed in AppVeyor artifact OpenSim-e49a8bde-2018-05-31-win64.exe:

![](https://user-images.githubusercontent.com/4203505/40805315-b92d265e-64d2-11e8-843b-c2c0bebaef24.png)

We're currently using RGB(232, 216, 200) for bones---though we've changed the color of one of the lights as well.
